File: Aggressive Cows - GFG/aggressive-cows.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        int t;
        t = sc.nextInt();
        while (t-- > 0) {

            int n;
            n = sc.nextInt();

            int k;
            k = sc.nextInt();

            int[] v = new int[n];
            for (int i = 0; i < n; i++) v[i] = sc.nextInt();

            Solution obj = new Solution();
            int res = obj.solve(n, k, v);

            System.out.println(res);
        }
    }
}

// } Driver Code Ends


// User function Template for Java
class Solution {
    public static boolean isPossible(int n,int k,int mid,int []stalls) {
		int prevcow = stalls[0];
		int cowcount = 1;
		
		for(int i =0;i<n;i++) {
			if(stalls[i] - prevcow>=mid) {
				cowcount++;
				prevcow = stalls[i];
				if(cowcount==k) {
					return true;
				}
			}
		}
		return false;
	}

    public static int solve(int n, int k, int[] a) {
		Arrays.sort(a);
		int s = 1;
		int e = a[n-1] - a[0];
		int ans = 0;
		int mid = s + (e-s)/2;
		while(s<=e) {
			if(isPossible(n,k,mid,a)) {
				ans = mid;
				s = mid+1;
			}
			else {
				e = mid -1;
			}
			mid = s + (e-s)/2;
		}
		return ans;
        
    }
}

File: Aggressive Cows - GFG/aggressive-cows.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
public:

    bool isPossible(int n,int k,vector<int> &stalls, int mid){
        int countcows = 1;
        int prevcow = stalls[0];
        int dis = 0;
        for(int i =1;i<n;i++){

            if (stalls[i] - prevcow >=mid){
                countcows++;
                if(countcows==k){
                    return true;
                }
                prevcow = stalls[i];
            }
        
        }
        return false;
    }
    int solve(int n, int k, vector<int> &stalls) {
    
        // Write your code here
        sort(stalls.begin(),stalls.end());
        int s = 1;
        int e = stalls[n-1] - stalls[0];
        
        int mid = s + (e-s)/2;
        int ans = 0;
        while(s<=e){
            if(isPossible(n,k,stalls,mid)){
                ans = mid;
                s = mid +1;
            }
            else{
               e = mid -1; 
            }
            mid = s + (e-s)/2;
        }
        return ans;
    }
};

//{ Driver Code Starts.

int main() {
    int t = 1;
    cin >> t;

    // freopen ("output_gfg.txt", "w", stdout);

    while (t--) {
        // Input

        int n, k;
        cin >> n >> k;

        vector<int> stalls(n);
        for (int i = 0; i < n; i++) {
            cin >> stalls[i];
        }
        // char ch;
        // cin >> ch;

        Solution obj;
        cout << obj.solve(n, k, stalls) << endl;

        // cout << "~\n";
    }
    // fclose(stdout);

    return 0;
}
// } Driver Code Ends